Add a unique suffix to kw environments

geraldino2 opened the pull request:
kw currently stores environment data in `${XDG_CACHE_HOME:-"${HOME}/.cache"}/${KWORKFLOW}/${ENV_DIR}/${env_name}`, and allows users to create non-distinct environment names in different work trees. When the new environment name isn't unique, data gets overwritten.

This PR implements a suffix to `env_name`: `$PWD`, encoded using base64 (available in Arch/Debian/Fedora through `coreutils`, already in `*.dependencies`).

It should fix issue #1185.

rodrigosiqueira commented on src/lib/kwlib.sh:
```diff
@@ -695,6 +695,25 @@ function get_current_env_name()
+# Return:
+# Returns 0 and prints ${current_env}-base64(${PWD}) if env_name is provided.
+# Returns 1 if env_name is not provided.
+get_env_name_encoded_with_pwd() {
```
This is not the codestyle adopted by kw, please, read the below page and review your entire PR.

 https://kworkflow.org/content/codingstyle.html

rodrigosiqueira commented on src/lib/kwlib.sh:
```diff
@@ -695,6 +695,25 @@ function get_current_env_name()
+  local env_name="$1"
+
+  if [[ -z "$env_name" ]]; then
+    return 1
```
Return 22 makes more sense here (EINVAL).

rodrigosiqueira commented on src/lib/kwlib.sh:
```diff
@@ -695,6 +695,25 @@ function get_current_env_name()
+    return 1
+  fi
+
+  local encoded_pwd=$(printf '%s' "$PWD" | base64)
```
move the `local encoded_pwd` declaration to the top of the function.

rodrigosiqueira commented on src/lib/kwlib.sh:
```diff
@@ -695,6 +695,25 @@ function get_current_env_name()
+    return 1
+  fi
+
+  local encoded_pwd=$(printf '%s' "$PWD" | base64)
```
Regarding base64, this seems excessive for this problem.  Additionally, I think base64 can return a multiple-line output, which can be a problem here.  I suppose if you want to use it, you need to add a wrapper to ensure the output is not displayed on multiple lines.

However, a simple md5sum would be better here.

geraldino2 replied:
Nice catch, I wasn't used to this multiple-line output. base64 on MacOS doesn't have it, but coreutils has, indeed, a default line limit of 76 chars. I'll disable line wrapping.

Is there any reason to use md5sum? base64 seems to do well: reversible and unique.

rodrigosiqueira commented on src/lib/kwlib.sh:
```diff
@@ -695,6 +695,25 @@ function get_current_env_name()
+
+  local encoded_pwd=$(printf '%s' "$PWD" | base64)
+
+  printf '%s-%s' "$env_name" "$encoded_pwd"
```
I don't think you need to add the env_name; you can use the encoded_pwd as a subfolder, and inside it, you keep the env name. In other words, to solve this problem, you just create a new subfolder that is the encoded version of the path.

geraldino2 replied:
Good idea.

rodrigosiqueira commented on src/lib/kwlib.sh:
The unit test should be part of this commit.

rodrigosiqueira commented on src/lib/kwlib.sh:
```diff
@@ -714,6 +714,25 @@ get_env_name_encoded_with_pwd() {
+# Return:
+# Returns 0 if successful, 1 if input is empty or malformed.
+# Prints the env_name (everything before the last hyphen) if successful.
+decode_env_name_with_pwd() {
```
from my comments in the first commit, I'm not convinced that we need this function.

rodrigosiqueira commented on src/kw_env.sh:
```diff
@@ -109,13 +109,14 @@ function validate_env_before_switch()
   flag=${flag:-'SILENT'}
 
-  if [[ ! -d "${local_kw_configs}/${ENV_DIR}/${target_env}" ]]; then
+  if [[ ! -d "${local_kw_configs}/${ENV_DIR}/${encoded_target_env}" ]]; then
```
iirc, the original issue reported by Melissa was about issues with the env collision in the cache, not in the local config. In this sense, you should focus on things that touch `KW_CACHE_DIR`, not local `.kw` config files. In other words, you should not change these parts. Am I missing something?

geraldino2 replied:
You are correct, changes on `.kw` aren't necessary. The goal was to have consistent naming between local/global contexts. However, as you suggested, I'm going to use `base64(path)/env_name` in `KW_CACHE_DIR`, so I'll revert changes in `.kw`.

rodrigosiqueira commented on src/kw_env.sh:
```diff
@@ -213,42 +217,42 @@ function create_new_env()
   # If we already have an env, we should copy the config file from it.
   if [[ "$ret" == 0 ]]; then
-    cmd="cp ${cache_build_path}/${ENV_DIR}/${current_env_name}/.config ${cache_build_path}/${ENV_DIR}/${env_name}/.config"
+    cmd="cp ${cache_build_path}/${ENV_DIR}/${current_env_name}/.config ${cache_build_path}/${ENV_DIR}/${encoded_env_name}/.config"
```
As I commented before, it would be better if we just create a new folder with the encoded name (without the env name), and inside it we create the env name. For example, suppose the encoded_pwd is `845d87776ab0bcd0c17f051d264f1cf2`, and the env names are `RPI64` and `x86_64`, we would have something like this:

`${cache_build_path}/${ENV_DIR}/${encoded_pwd}/${current_env_name}/`

rodrigosiqueira commented on tests/unit/kw_env_test.sh:
unit test should be in the same commit that introduces the tested modifications.

rodrigosiqueira reviewed (changes requested):
Hi,
First of all, thanks for your PR. I think you are in the right direction, but you need to improve some things:
1. Organize your commits. Usually, PR reviews happen on a commit basis, which means that we review from the first commit to the last one. Every commit should be self-contained, and tests should be introduced in the same commit as the feature. For example, if you added function X in the first commit, then you should add the test for X in the same commit. Also, if one commit improves something introduced in a previous commit that has not been merged yet, then you should squash your new commit into the previous one. Anyway, please take the time to improve your commits.
2. Keep in mind that envs have two types of metadata: local configs (.kw) and global folders (.cache). For this problem afaik, you just need to focus on the global part.
3. Finally, take some time to read the kw codestyle guideline.

geraldino2 committed:
src: lib: kwlib: Introduce a function to encode $PWD

This commit introduces a function that returns the pwd string encoded in base64.

Signed-off-by: Gabriel Geraldino <gabrielgeraldinosouza@gmail.com>
Co-developed-by: Nicolas Borsari <nicborsari@usp.br>
Signed-off-by: Nicolas Borsari <nicborsari@usp.br>

geraldino2 committed:
src: kw_env: Adjust functions to use a encoded PWD as a cache_build_path

This commit implements get_encoded_pwd function in the global context
for solving duplicated env_names in different kernel trees. For every
env inside a tree, it will be created a folder in the global context
with its encoded pwd. The proper env_name folder will be created inside
it.

Signed-off-by: Gabriel Geraldino <gabrielgeraldinosouza@gmail.com>
Co-developed-by: Nicolas Borsari <nicborsari@usp.br>
Signed-off-by: Nicolas Borsari <nicborsari@usp.br>

geraldino2 commented:
Hi, thanks for the feedback. The PR is looking better now. I've adjusted the code to focus only on the global part and organized the commits. About the codestyle, I think the problem was originally addressed in the `chore` commits; I don't see any other issue.

Do you mind taking a look again?

rodrigosiqueira reviewed (changes requested):
Hi,

Thanks a lot for your update. I think you are close to finishing it. I added some minor comments, but there is one final issue to handle.

One thing to keep in mind is that we already have users who use the kw env feature. In this sense, you also need to provide one mid-step to migrate the env path to the new one. You can do it in a new commit. To validate what I mean, you can try the following:
1. Outside your branch, create two or three new envs. Ensure that you add a .config file to each one; in other words, create a realistic environment.
2. Move back to your env, and try to create a new env. You will encounter your first error.
My suggestion is:
1. Create a function named `migrate_old_env_to_base64` (or something better). Add a comment in this function to indicate that we can remove it in the future.
2. This function should check the cache folder to see if the expected level of directories is present, i.e., your modification introduced a folder new level, If it is not found, you can assume it is the old approach.
3. Add this function at the beginning of every env function to ensure that the migration happens as soon as possible.

Thanks

ncbor committed:
src: kw_env: Migrate existing envs to a new path

kw env was unable to create environments with the same name on
different kernel trees. Environments are now being created on an
individual folder for each tree, inside ${KW_CACHE_DIR}. However,
this change doesn't have backward compatibility.

This commit introduces a migration step before kw commands.

Signed-off-by: Gabriel Geraldino <gabrielgeraldinosouza@gmail.com>
Co-developed-by: Nicolas Borsari <nicborsari@usp.br>
Signed-off-by: Nicolas Borsari <nicborsari@usp.br>

geraldino2 commented:
Hi,

I created `migrate_old_envs_to_base64`, it's being called on `env_main`. Unless someone is in the case reported in #1185, migration should go well.

By the way, since I'm already modifying these files, do you want me to add the changes below to the current PR?
- src/kw_env.sh:list_env_available_envs: array is loaded with `find "${local_kw_configs}/${ENV_DIR}" -maxdepth 1` and its size is compared to `0`. If `find` has no `mindepth`, the base path is also printed. It's an easy bug that can be fixed by adding `mindepth 1`. The second `no_env_message` on `list_env_available_envs` is unreachable right now;
- tests/unit/kw_env_test.sh: `$LINENO` is being used instead of `${LINENO}`.

Do you mind taking a look again?

Thanks

rodrigosiqueira commented:
> Hi,
> 
> I created `migrate_old_envs_to_base64`, it's being called on `env_main`. Unless someone is in the case reported in #1185, migration should go well.
> 
> By the way, since I'm already modifying these files, do you want me to add the changes below to the current PR?
> 
>     * src/kw_env.sh:list_env_available_envs: array is loaded with `find "${local_kw_configs}/${ENV_DIR}" -maxdepth 1` and its size is compared to `0`. If `find` has no `mindepth`, the base path is also printed. It's an easy bug that can be fixed by adding `mindepth 1`. The second `no_env_message` on `list_env_available_envs` is unreachable right now;
> 
>     * tests/unit/kw_env_test.sh: `$LINENO` is being used instead of `${LINENO}`.
> 
> 
> Do you mind taking a look again?
> 
> Thanks

yes, please, I appreciate that. Add all of them in their own commit.

rodrigosiqueira commented on tests/unit/kw_env_test.sh:
```diff
@@ -346,4 +346,95 @@ function test_create_and_destroy_env()
+  assertFalse "(${LINENO}) Old env2 should be moved" '[[ -d "${old_env_path}/env2" ]]'
+
+  # Verify config content is preserved
+  local content1=$(cat "${new_env_path}/env1/build.config")
+  local content2=$(cat "${new_env_path}/env2/build.config")
```
https://kworkflow.org/content/codingstyle.html#command-substitution-and-arithmetic-expression

Also, move the local declaration to the beginning of the function.

rodrigosiqueira reviewed (changes requested):
Hi,
That is a very nice update! I tested it locally, and it looks like it is working as expected. I added some minor comments; I'm looking forward to the new version.

Thanks a lot.

rodrigosiqueira commented:
Hi,
I made some changes myself, and merged it into the unstable branch. Thanks.
